module/os/linux/zfs/abd.c

```diff
@@ -207,6 +207,15 @@ static abd_stats_t abd_stats = {
 #define	abd_for_each_sg(abd, sg, n, i)	\
 	for_each_sg(ABD_SCATTER(abd).abd_sgl, sg, n, i)
 
+
+#define	ABD_MULTI(abd)		(abd->abd_u.abd_multi)
+#define	abd_is_multi(abd)	((abd->abd_flags & ABD_FLAG_MULTI_LIST) != 0)
+
+typedef struct abd_link {
+	abd_t 		*link_abd;
+	list_node_t link_node;
+} abd_link_t;
```

Comment:
I definitely agree that putting the list_node_t inside of the abd_t struct is a much simpler approach. However, as you stated, I think leaving it as is would allow the same abd to be in multiple “multi” abd’s. I am fairly certain this can in fact happen in the Direct IO code path that is in our other PR.

We construct “multi” abd’s when the request is not aligned perfectly. In turn that same abd can be in the vdev aggregate function as well. Just in a forward thinking approach, with the Direct IO work, I think we would want to leave the list_node_t in a separate struct.

Reply:
I would like to investigate if we really need the unaligned handling in the other PR.  I would guess that we don't really care about performance of unaligned directios, so perhaps we can fall back on a simpler implementation (even if it is less efficient).

Reply:
We should do the investigation, but my guess is that there are compelling use cases for supporting "unaligned" directIO. Note that the alignment constraint is against the block size. E.g., if we don't have a full block read, we construct a full block by using a multi-abd. So a use case example would be a database that is construct using an 8K ZFS block size is then read using 4K directIO random reads. I think we will see a significant performance difference if we fall back to doing a data copy.

Reply:
In our performance investigations with recordsize=8k, I haven't seen the per-byte costs like bcopy() be significant (even with compression=lz4 and checksum=edonr, both more expensive than the defaults).  CPU time is almost always dominated by the per-block costs (e.g. creating dbufs, arc bufs, etc, and tearing them down).  One of my concerns is that allocating/freeing the abd_link_t would add to the per-block costs.

That said, I agree that in some cases, partial-block directio reads could benefit from the directio semantics.  Bringing us back to this PR, could you help me understand how that leads to an abd_t being part of more than one multi-abd?  I get that for the partial-block directio read, we create a multi-abd where the children are the user's buffer plus a dummy buffer that we will later throw away.  Does the user's abd (as opposed to the first multi-abd) then become part of a different multi-abd?

Reply:
Heh, I've actually been thinking a bit about this. I don't think there is a place where we will try to use an abd_t in more than one multi-abd. I don't remember when/why the abd_link_t was introduced (although I'm pretty sure I introduced it). It may have been a misguided attempt to avoid expanding the abd_t struct size.

Reply:
Cool, it sounds like we can just go ahead and place the list_node_t inside of the abd_t's struct then.

Reply:
@mmaybee, are we sure that abd_t can not be in multiple multi-abd's at once?

I just updated the code to have the abd_t contain the list_node_t and was creating multi-abd's chaining them together with the list_node_t member variable. However, we now are getting all kinds of failed ASSERT's in abd_verify(). In particular we alway seem to fail the following:
ASSERT3U(abd->abd_size, >, 0)

Reply:
When adding an abd to a multi-abd, you should be able to assert that it isn't already in a multi-abd.  Something like `ASSERT(!list_link_active(&abd->abd_multi_link))`

Reply:
@ahrens, I added an ASSERT for !list_link_active() to abd_add_child() and that is now where the failure is happening. Thank you for pointing this out! It does seem that an abd_t can be present in multiple multi-abd's.

Reply:
In what scenario do we add one abd_t to multiple multi-abd's?

Reply:
@mmaybee are in the process of hunting this down at the moment.

Reply:
So we did wind up moving the list link inside of the ABD struct itself and addressed the situations where a ABD can be in multiple Gang ABDs.